app/models/financialDetails/CreditType.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package models.financialDetails
+
+sealed trait CreditType {
+  val key: String
+}
+
+case object MfaCreditType extends CreditType {
+  override val key = "mfa"
+}
+
+case object CutOverCreditType extends CreditType {
+  override val key = "cutOver"
+}
+
+case object BalancingChargeCreditType extends CreditType {
+  override val key = "balancingCharge"
+}
+
+case object RepaymentInterest extends CreditType {
+  override val key = "repaymentInterest"
+}
+
+case object PaymentType extends CreditType {
+  override val key = "payment"
+}
+
+case object Repayment extends CreditType {
+  override val key = "refund"
+}
+
+object CreditType {
+
+  // values come from EPID #1138
+  private val cutOver = "6110"
+  private val balancingCharge = "4905"
+  private val repaymentInterest = "6020"
+  private val mfaCredit = Range.inclusive(4004, 4025)
+    .filterNot(_ == 4010).filterNot(_ == 4020).map(_.toString)
+    .toList
+  private val payment = List("4920", "4930")
+
+  def apply(mainTransaction: String): Option[CreditType] = {
```

Comment:
its looks like this is more "unapply" semantic rather than "appy"

Reply:
I'd disagree - it operates more like a constructor? It takes the mainTransaction as an argument and returns a subtype of CreditType?

Reply:
apply is for constructors returning plain type; here is for unapply:
```
The return type of an unapply should be chosen as follows:

If it is just a test, return a Boolean. For instance case even().
If it returns a single sub-value of type T, return an Option[T].
If you want to return several sub-values T1,...,Tn, group them in an optional tuple Option[(T1,...,Tn)].
```

Reply:
https://docs.scala-lang.org/scala3/reference/other-new-features/creator-applications.html
apply is a shortcut to drop usage of new, nothing else

Reply:
The link is just saying that the automatically generated apply methods on case classes have been extended to concrete classes - it's not saying apply cannot be used in this context.

I think apply being used to implement a Factory pattern is valid, concise and readable 🙂

https://www.oreilly.com/library/view/scala-cookbook/9781449340292/ch06s10.html

Reply:
method logic clearly contains validation logic, I really dont think this is something to do with how apply is intended to be used;

Reply:
Shall we just rename this method, then? Something like CreditType.parse(mainTransaction) or CreditType.fromCode(mainTransaction)?

Reply:
* you can keep it as is, there is not harm in this, its just not advisable practice in my opinion;
* renaming is also a good option, as this seem to be like type conversion not sure about validation;
* also I was thinking next approach:

mainTransactionString => Transaction
def apply(mainTransaction: Transaction): CreditType = ???

Reply:
I think renaming is the best option at the moment. It would be good to look into using specific types and validation for a lot of fields throughout (using validators when parsing objects from the JSON would be the usual way), but removing the Option[CreditType] return type would require some thought - there are valid mainTransaction codes that we don't currently support on these pages.